src/Utilities/penalty_relaxation.jl

```diff
@@ -0,0 +1,239 @@
+true
+```
+"""
+mutable struct PenaltyRelaxation{T}
+    default::Union{Nothing,T}
+    penalties::Dict{MOI.ConstraintIndex,T}
+
+    function PenaltyRelaxation(
+        p::Dict{MOI.ConstraintIndex,T};
+        default::Union{Nothing,T} = one(T),
+    ) where {T}
+        return new{T}(default, p)
+    end
+end
+
+function PenaltyRelaxation(; kwargs...)
+    return PenaltyRelaxation(Dict{MOI.ConstraintIndex,Float64}(); kwargs...)
+end
+
+function PenaltyRelaxation(
+    d::Dict{<:MOI.ConstraintIndex,T};
+    kwargs...,
+) where {T}
+    return PenaltyRelaxation(convert(Dict{MOI.ConstraintIndex,T}, d); kwargs...)
+end
+
+function MOI.modify(model::MOI.ModelLike, relax::PenaltyRelaxation{T}) where {T}
+    sense = MOI.get(model, MOI.ObjectiveSense())
+    if sense == MOI.FEASIBILITY_SENSE
+        MOI.set(model, MOI.ObjectiveSense(), MOI.MIN_SENSE)
+        f = zero(MOI.ScalarAffineFunction{T})
+        MOI.set(model, MOI.ObjectiveFunction{typeof(f)}(), f)
+    end
+    map = Dict{MOI.ConstraintIndex,MOI.ScalarAffineFunction{T}}()
+    for (F, S) in MOI.get(model, MOI.ListOfConstraintTypesPresent())
+        if MOI.supports(model, relax, MOI.ConstraintIndex{F,S})
+            _modify_penalty_relaxation(map, model, relax, F, S)
+        else
+            @warn("Skipping PenaltyRelaxation of constraints of type $F-in-$S")
+        end
+    end
+    return map
+end
+
+function _modify_penalty_relaxation(
+    map::Dict{MOI.ConstraintIndex,MOI.ScalarAffineFunction{T}},
+    model::MOI.ModelLike,
+    relax::PenaltyRelaxation,
+    ::Type{F},
+    ::Type{S},
+) where {T,F,S}
+    for ci in MOI.get(model, MOI.ListOfConstraintIndices{F,S}())
+        if relax.default !== nothing || haskey(relax.penalties, ci)
+            f = MOI.modify(model, ci, relax)
+            if f !== nothing
+                map[ci] = f
+            end
+        end
+    end
+    return map
+end
+
+function MOI.supports(
+    ::MOI.ModelLike,
+    ::PenaltyRelaxation{T},
+    ::Type{<:MOI.ConstraintIndex{F,<:MOI.AbstractScalarSet}},
+) where {T,F<:Union{MOI.ScalarAffineFunction{T},MOI.ScalarQuadraticFunction{T}}}
+    return true
+end
+
+function MOI.supports(
+    ::MOI.ModelLike,
+    ::PenaltyRelaxation,
+    ::Type{<:MOI.ConstraintIndex},
+)
+    return false
+end
+
+function MOI.modify(
+    model::MOI.ModelLike,
+    ci::MOI.ConstraintIndex{F,<:MOI.AbstractScalarSet},
+    relax::PenaltyRelaxation{T},
+) where {T,F<:Union{MOI.ScalarAffineFunction{T},MOI.ScalarQuadraticFunction{T}}}
+    y = MOI.add_variable(model)
+    z = MOI.add_variable(model)
+    MOI.add_constraint(model, y, MOI.GreaterThan(zero(T)))
+    MOI.add_constraint(model, z, MOI.GreaterThan(zero(T)))
+    MOI.modify(model, ci, MOI.ScalarCoefficientChange(y, one(T)))
+    MOI.modify(model, ci, MOI.ScalarCoefficientChange(z, -one(T)))
+    sense = MOI.get(model, MOI.ObjectiveSense())
+    scale = sense == MOI.MIN_SENSE ? one(T) : -one(T)
+    a = scale * get(relax.penalties, ci, relax.default)
```

Comment:
It's a bit weird that a constraint would receive a dictionary and not its value directly. Maybe we should have a different attribute for `modify` with a constraint index that just contains a value. `ScalarCoefficientChange` contains just the value, not a dictionary so it would be inconsistent to have a dictionary here.

Reply:
I added `ScalarPenaltyRelaxation`. Does that help?